### PR TITLE
Make conversion from `hex` to `bin`

### PR DESCRIPTION
**Pull request by FedericoBruzzone**
Signed-off-by: Federico Bruzzone <federico.bruzzone.i@gmail.com>

### CODE REVIEW

**Comment by FedericoBruzzone** [comment]
Reference to issue #35

**Comment by FedericoBruzzone** [comment]
I commented `cartridge_header.rs` with the binary conversion.

I don't know if it's better to keep it in hex, let me know.

File: src/arm7tdmi.rs [hunk 1]
```diff
@@ -239,18 +240,18 @@ mod tests {
         let mut opcode: u32 = 0b1110_0011_1010_0000_0000_0000_0000_0000;
 
         // bits [11-8] are ROR-Shift applied to nn
-        let is = opcode & 0x00_00_0F_00;
+        let is = opcode & 0b0000_0000_0000_0000_0000_1111_0000_0000;
 
         // MOV Rx,x
         let mut cpu = Arm7tdmi::new(vec![]);
-        for rx in 0..=0xF {
+        for rx in 0..=0b1111 {
```
Comment (guerinoni) [inline src/arm7tdmi.rs:247]: Let's keep this as hex

File: src/cartridge_header.rs [hunk 2]
```diff
@@ -63,20 +63,23 @@ impl CartridgeHeader {
     }
 
     // 32bit ARM branch opcode, eg. "B rom_start"
+    // 0b0000_0000_0000..=0b0000_0000_0011
```
Comment (guerinoni) [inline src/cartridge_header.rs:66]: Not relevant, is enough readable

File: src/cartridge_header.rs [hunk 3]
```diff
@@ -63,20 +63,23 @@ impl CartridgeHeader {
...
     fn extract_rom_entry_point(data: &[u8]) -> [u8; 4] {
         data[0x000..=0x003]
             .try_into()
             .expect("extracting rom entry point")
     }
 
     // compressed bitmap, required!
+    // 0b0000_0000_0100..=0b0000_1001_1111
```
Comment (guerinoni) [inline src/cartridge_header.rs:74]: Not relevant, is enough readable

File: src/cartridge_header.rs [hunk 4]
```diff
@@ -63,20 +63,23 @@ impl CartridgeHeader {
...
     fn extract_nintendo_logo(data: &[u8]) -> [u8; 156] {
         data[0x004..=0x09F]
             .try_into()
             .expect("extracting nintendo logo")
     }
 
     // uppercase ascii, max 12 characters
+    // 0b0000_1010_0000..=0b0000_1010_1011
```
Comment (guerinoni) [inline src/cartridge_header.rs:82]: Not relevant, is enough readable

File: src/cartridge_header.rs [hunk 5]
```diff
@@ -86,6 +89,7 @@ impl CartridgeHeader {
     }
 
     // uppercase ascii, 4 characters
+    // 0b0000_1010_1100..=0b0000_1010_1111
```
Comment (guerinoni) [inline src/cartridge_header.rs:92]: Not relevant, is enough readable

File: src/cartridge_header.rs [hunk 6]
```diff
@@ -95,6 +99,7 @@ impl CartridgeHeader {
     }
 
     // uppercase ascii, 2 characters
+    // 0b0000_1011_0000..=0b0000_1011_0001
```
Comment (guerinoni) [inline src/cartridge_header.rs:102]: Not relevant, is enough readable

File: src/cartridge_header.rs [hunk 7]
```diff
@@ -212,11 +229,11 @@ impl CartridgeHeader {
 }
 
 fn verify_checksum(data: &[u8]) -> Result<(), ()> {
-    let checksum_expected = data[0xBD];
-    let checksum = data[0xA0..=0xBC]
+    let checksum_expected = data[0xBD]; // 0b0000_1011_1101
+    let checksum = data[0xA0..=0xBC] // 0b0000_1010_0000..=0b0000_1011_1100
```
Comment (guerinoni) [inline src/cartridge_header.rs:233]: Not relevant, is enough readable

File: src/cartridge_header.rs [hunk 8]
```diff
@@ -194,11 +209,13 @@ impl CartridgeHeader {
     }
 
     // seems to be unused
+    // 0b0000_1100_0110..=0b0000_1100_1111
     fn extract_not_used(data: &[u8]) -> [u8; 26] {
         data[0x0C6..=0x0DF].try_into().expect("extracting not used")
     }
 
     // 32bit ARM branch opcode, eg. "B joy_start"
+    // 0b0000_1110_0000..=0b0000_1110_0011
```
Comment (guerinoni) [inline src/cartridge_header.rs:218]: Not relevant, is enough readable

File: src/cartridge_header.rs [hunk 9]
```diff
@@ -194,11 +209,13 @@ impl CartridgeHeader {
     }
 
     // seems to be unused
+    // 0b0000_1100_0110..=0b0000_1100_1111
```
Comment (guerinoni) [inline src/cartridge_header.rs:212]: Not relevant, is enough readable

File: src/cartridge_header.rs [hunk 10]
```diff
@@ -178,6 +192,7 @@ impl CartridgeHeader {
     }
 
     // init as 00h - BIOS overwrites this value!
+    // 0b0000_1100_0101..=0b0000_1100_0101
```
Comment (guerinoni) [inline src/cartridge_header.rs:195]: Not relevant, is enough readable

File: src/cartridge_header.rs [hunk 11]
```diff
@@ -163,6 +176,7 @@ impl CartridgeHeader {
     }
 
     // init as 00h - BIOS overwrites this value!
+    // 0b0000_1100_0100..=0b0000_1100_0100
```
Comment (guerinoni) [inline src/cartridge_header.rs:179]: Not relevant, is enough readable

File: src/cartridge_header.rs [hunk 12]
```diff
@@ -147,13 +158,15 @@ impl CartridgeHeader {
...
     // should be zero filled
+    // 0b0000_1011_1110..=0b0000_1011_1111
     fn extract_reserved_area_2(data: &[u8]) -> [u8; 2] {
         data[0x0BE..=0x0BF]
             .try_into()
             .expect("extracting reserved area 2")
     }
 
     // 32bit ARM branch opcode, eg. "B ram_start"
+    // 0b0000_1100_0000..=0b0000_1100_0011
```
Comment (guerinoni) [inline src/cartridge_header.rs:169]: Not relevant, is enough readable

File: src/cartridge_header.rs [hunk 13]
```diff
@@ -147,13 +158,15 @@ impl CartridgeHeader {
     }
 
     // should be zero filled
+    // 0b0000_1011_1110..=0b0000_1011_1111
```
Comment (guerinoni) [inline src/cartridge_header.rs:161]: Not relevant, is enough readable

File: src/cartridge_header.rs [hunk 14]
```diff
@@ -104,41 +109,47 @@ impl CartridgeHeader {
...
     // usually 00h
+    // 0b0000_1011_1100..=0b0000_1011_1100
     fn extract_software_version(data: &[u8]) -> [u8; 1] {
         data[0x0BC..=0x0BC]
             .try_into()
             .expect("extracting software version")
     }
 
     // header checksum, required!
+    // 0b0000_1011_1101..=0b0000_1011_1101
```
Comment (guerinoni) [inline src/cartridge_header.rs:152]: Not relevant, is enough readable

File: src/cartridge_header.rs [hunk 15]
```diff
@@ -104,41 +109,47 @@ impl CartridgeHeader {
...
     // should be zero filled
+    // 0b0000_1011_0101..=0b0000_1011_1011
     fn extract_reserved_area_1(data: &[u8]) -> [u8; 7] {
         data[0x0B5..=0x0BB]
             .try_into()
             .expect("extracting reserved area 1")
     }
 
     // usually 00h
+    // 0b0000_1011_1100..=0b0000_1011_1100
```
Comment (guerinoni) [inline src/cartridge_header.rs:144]: Not relevant, is enough readable

File: src/cartridge_header.rs [hunk 16]
```diff
@@ -104,41 +109,47 @@ impl CartridgeHeader {
...
     // usually 00h (bit7=DACS/debug related)
+    // 0b0000_1011_0100..=0b0000_1011_0100
     fn extract_device_type(data: &[u8]) -> [u8; 1] {
         data[0x0B4..=0x0B4]
             .try_into()
             .expect("extracting device type")
     }
 
     // should be zero filled
+    // 0b0000_1011_0101..=0b0000_1011_1011
```
Comment (guerinoni) [inline src/cartridge_header.rs:136]: Not relevant, is enough readable

File: src/cartridge_header.rs [hunk 17]
```diff
@@ -104,41 +109,47 @@ impl CartridgeHeader {
     }
 
     // must be 96h, required!
+    // 0b0000_1011_0010..=0b0000_1011_0010
```
Comment (guerinoni) [inline src/cartridge_header.rs:112]: Not relevant, is enough readable

File: src/cartridge_header.rs [hunk 18]
```diff
@@ -104,41 +109,47 @@ impl CartridgeHeader {
...
     fn extract_fixed_value(data: &[u8]) -> [u8; 1] {
         data[0x0B2..=0x0B2]
             .try_into()
             .expect("extracting fixed value")
     }
 
     // 00h for current GBA models
+    // 0b0000_1011_0011..=0b0000_1011_0011
```
Comment (guerinoni) [inline src/cartridge_header.rs:120]: Not relevant, is enough readable

File: src/cartridge_header.rs [hunk 19]
```diff
@@ -104,41 +109,47 @@ impl CartridgeHeader {
...
     fn extract_main_unit_code(data: &[u8]) -> [u8; 1] {
         data[0x0B3..=0x0B3]
             .try_into()
             .expect("extracting main unit code")
     }
 
     // usually 00h (bit7=DACS/debug related)
+    // 0b0000_1011_0100..=0b0000_1011_0100
```
Comment (guerinoni) [inline src/cartridge_header.rs:128]: Not relevant, is enough readable

**Review by guerinoni — CHANGES_REQUESTED** [review]

File: src/cpsr.rs [hunk 20]
```diff
@@ -15,7 +15,7 @@ impl Cpsr {
     }
 
     fn signed(&self) -> bool {
-        self.0 & 0x8000 != 0
+        self.0 & 0b1000_0000_0000_0000 != 0
```
Comment (FedericoBruzzone) [inline src/cpsr.rs:18]: I have only converted to binary `0x80_00` , but it shouldn't be `0x80_00_00_00`?

`sign_flag` is 31st bit ([Documentation](https://rust-console.github.io/gbatek-gbaonly/#armcpuflagsconditionfieldcond))
Immagine 2022-09-09 002431

Comment (guerinoni) [reply src/cpsr.rs:18]: Can you add a test that should verify this error? If this is true, open a new PR because is important

Comment (FedericoBruzzone) [reply src/cpsr.rs:18]: I'll try, I'm still learning rust, I've been studying it for 5 days.

**Comment by FedericoBruzzone** [comment]
I opened a new PR.
Do you want to keep this open to fix the error? (From 0xAAA to 0xAAAAAAAA)?

File: src/cpsr.rs (hunk 20 above)
Comment (FedericoBruzzone) [reply src/cpsr.rs:18]: I opened a new PR.
Do you want to keep this open to fix the error? (From 0xAAAA to 0xAAAAAAAA)?

**Pull request closed without merging**